pkg/scheduler/internal/queue/scheduling_queue.go

```diff
@@ -772,10 +772,24 @@ func (npm *nominatedPodMap) delete(p *v1.Pod) {
 }
 
 func (npm *nominatedPodMap) update(oldPod, newPod *v1.Pod) {
+	// In some cases, an Update event with no "NominatedNode" present is received right
+	// after a node("NominatedNode") is reserved for this pod in memory.
+	// In this case, we need to keep reserving the NominatedNode when updating the pod pointer.
+	nodeName := ""
+	// We won't fall into below `if` block if the Update event represents:
+	// (1) NominatedNode info is added
+	// (2) NominatedNode info is updated
+	// (3) NominatedNode info is removed
+	if NominatedNodeName(oldPod) == "" && NominatedNodeName(newPod) == "" {
+		if nnn, ok := npm.nominatedPodToNode[oldPod.UID]; ok {
```

Comment:
A neat way is `nodeName = npm.nominatedPodToNode[oldPod.UID]`. If oldPod.UID doesn't exist, a nil string is returned.

Reply:
This change makes the logic more robust, but `nodeName` should not be initialized with empty string. It should be initialized by `NominatedNodeName(newPod)`. Otherwise, all usual updates would remove nominated pods.

Reply:
Initilizing `nodeName` as empty string should be good b/c in nominatedPodMap.add(), it checks if incoming nodeName is empty; if it is, it calls NominatedNodeName(p):

https://github.com/kubernetes/kubernetes/blob/314264aeafb94d4948a3f117bcc4382eba10dede/pkg/scheduler/internal/queue/scheduling_queue.go#L735-L746

Reply:
You are right.